engine/src/it/java/org/pentaho/di/trans/steps/prioritizestreams/PrioritizeStreamsExecutionIT.java

```diff
@@ -0,0 +1,280 @@
+    }
+
+    Assert.assertEquals( "Output stream collect all rows: " + ps.toString(), 9, ps.output.size() );
+    Assert.assertEquals( "Stream output respect priority", code.toString(), ps.toString() );
+  }
+
+  private PrioritizeStreamsData getTestData( TestData code ) {
+    PrioritizeStreamsData retData = new PrioritizeStreamsData();
+
+    RowSet in1 = new BlockingRowSet( 4 );
+    RowSet in2 = new BlockingRowSet( 4 );
+    RowSet in3 = new BlockingRowSet( 4 );
+    Object[][] data1 = { { "a" }, { "a" }, { "a" } };
+    Object[][] data2 = { { "b" }, { "b" }, { "b" } };
+    Object[][] data3 = { { "c" }, { "c" }, { "c" } };
+    RowMetaInterface rmi = new RowMeta();
+
+    Runnable stream1 = this.getInputProduser( in1, rmi, data1 );
+    Runnable stream2 = this.getInputProduser( in2, rmi, data2 );
+    Runnable stream3 = this.getInputProduser( in3, rmi, data3 );
+
+    retData.outputRowMeta = rmi;
+
+    switch ( code ) {
+      case ABC: {
+        retData.rowSets = new RowSet[] { in1, in2, in3 };
+        break;
+      }
+      case BAC: {
+        retData.rowSets = new RowSet[] { in2, in1, in3 };
+        break;
+      }
+      case BCA: {
+        retData.rowSets = new RowSet[] { in2, in3, in1 };
+        break;
+      }
+      case CAB: {
+        retData.rowSets = new RowSet[] { in3, in1, in2 };
+        break;
+      }
+      case CBA: {
+        retData.rowSets = new RowSet[] { in3, in2, in1 };
+        break;
+      }
+      default: {
+        Assert.fail( "This test data does not supported: " + code.toString() );
+      }
+    }
+
+    retData.currentRowSet = retData.rowSets[0];
+
+    retData.stepnr = 0;
+    retData.stepnrs = 3;
+
+    service.execute( stream1 );
+    service.execute( stream2 );
+    service.execute( stream3 );
+
+    return retData;
+  }
+
+  private Runnable getInputProduser( final RowSet in, final RowMetaInterface rmi, final Object[][] data ) {
+    return new InputProducer( in, rmi, data );
+  }
+
+  /**
+   * This class simulates working delay before actual data will be available for
+   * input stream. Pay attention to KETTLE_ROWSET_GET_TIMEOUT value. Default
+   * value can be too small - and empty (even not started to fill) rowset will
+   * be treated as empty (returns 'null' for offer method call on blocking queue).
+   *
+   */
+  private class InputProducer implements Runnable {
+
+    private Queue<Object[]> data;
+    private RowSet rs;
+    private RowMetaInterface rmi;
+
+    private boolean first = true;
+
+    InputProducer( RowSet in, RowMetaInterface rmi, Object[][] data ) {
+      this.data = new LinkedList<Object[]>();
+      this.data.addAll( Arrays.asList( data ) );
+      this.rs = in;
+      this.rmi = rmi;
+    }
+
+    @Override
+    public void run() {
+      while ( data.peek() != null ) {
+        // first row always came with delay
+        if ( first ) {
+          first = false;
+          try {
+            Thread.sleep( 50 );
+          } catch ( InterruptedException e ) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: InterruptedException is ignored. This can delay thread shutdown and clear the thread’s interrupt status. Only code that implements a thread’s interruption policy can swallow an interruption request.

**Fix**: Rethrow the InterruptedException or reinterrupt the current thread using *Thread.currentThread().interrupt()* so that higher-level interrupt handlers can function correctly.
 If you are wrapping the InterruptedException inside a RuntimeException, call *Thread.currentThread().interrupt()* before throwing the RuntimeException.

Learn more about [interrupts](https://docs.oracle.com/javase/tutorial/essential/concurrency/interrupt.html) and [dealing with InterruptedException](https://www.ibm.com/developerworks/java/library/j-jtp05236/index.html?ca=drs-#N10187)